### ISSUE TITLE
For Michael Hirsch, re: "Numpy / OpenCV image BGR to RGB"

### ISSUE THREAD

**Issue by VideoPlayerCode** [scivision/pymap3d#30]
Hi Michael,

I was looking through all of your contact information on scivision.dev, and couldn't find any email address.

I'm just here to report that your article at https://www.scivision.dev/numpy-image-bgr-to-rgb/ is very harmful. There are plenty of people taking that advice and linking to that page, which is why I have to reach you. I used to believe that the same technique was good, but it turns out that there's _much_ more to BGR/RGB conversion than just modifying the "Numpy view" of the RAM storage. That's a "fake" conversion and doesn't actually work in real life.

Here's a full breakdown which explains what happens at a code-level in OpenCV when you give it such "faked" Numpy data:

https://answers.opencv.org/question/219040/fastest-way-to-convert-bgr-rgb-aka-do-not-use-numpy-magic-tricks/

The same is almost certainly true for matplotlib too.

With best regards,
Johnny

**Comment by scivision** [comment]
The discussion at the opencv.org link above notes that contiguous arrays should be passed to OpenCV to avoid copy time penalties each time. I think this is an important performance point that should be noted. Contiguous data operations are generally much faster whenever operating on data, whether within a language or when using APIs between languages like OpenCV.

**Comment by scivision** [comment]
I updated the page to reflect use of .copy.

**Comment by VideoPlayerCode** [comment]
@scivision Hi! :-) Yeah, using `.copy()` to ensure that data is contiguous is enough to ensure that no further OpenCV calls will need to make those slow, internally-copied contiguous images, and thanks for updating the page. However, note that _Numpy itself_ is ultra-slow at copying data whenever the strides have been modified to become non-contiguous (which is what `[...,::-1]` does; it messes up the strides). I wish Numpy would optimize and parallelize their code to make such copies fast, but they've stated that [they prefer not having any multithreaded code at all](https://github.com/numpy/numpy/issues/10291#issuecomment-486879785). So therefore you should always avoid stride-manipulation as much as possible. (Except when unavoidable, such as when one API needs `(colors, height, width)` order and another needs `(height, width, colors)`...)

Let's compare and see how slow Numpy is when copying stride-tweaked data:

**4K image, by changing Numpy stride and forcing it to copy, via `rgb = bgr[...,::-1].copy()`:**

`python -m timeit -s "import numpy as np; import cv2; bgr = np.zeros([2160,3840,3], np.uint8); bgr[:,:,2] = 255; bgr[:,:,1] = 100" "rgb = bgr[...,::-1].copy()"`

Result: **37.4** milliseconds per conversion.

**1920x1080 image, by changing Numpy stride and forcing it to copy, via `rgb = bgr[...,::-1].copy()`:**

`python -m timeit -s "import numpy as np; import cv2; bgr = np.zeros([1080,1920,3], np.uint8); bgr[:,:,2] = 255; bgr[:,:,1] = 100" "rgb = bgr[...,::-1].copy()"`

Result: **9.18** milliseconds per conversion.

---

And now the same color conversions, but directly giving OpenCV the contiguous BGR input (thus no internal OpenCV copy happens when you make the call since it's already contiguous), and letting `cv2.cvtColor` directly rearrange the color bytes of the existing contiguous RAM storage with its accelerated (multi-threaded) OpenCL code:

**4K image, through OpenCV's color converter, via `rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)`:**

`python -m timeit -s "import numpy as np; import cv2; bgr = np.zeros([2160,3840,3], np.uint8); bgr[:,:,2] = 255; bgr[:,:,1] = 100" "rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)"`

Result: **5.38** milliseconds per conversion. (6.95x faster)

**1920x1080 image, through OpenCV's color converter, via `rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)`:**

`python -m timeit -s "import numpy as np; import cv2; bgr = np.zeros([1080,1920,3], np.uint8); bgr[:,:,2] = 255; bgr[:,:,1] = 100" "rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)"`

Result: **1.52** milliseconds per conversion. (6.03x faster)

---

The guidelines at the page I linked earlier are as follows:

- If the data is already non-contiguous for some reason (such as another library's API that gave you a non-contiguous ndarray), then there's no way to avoid needing to trigger Numpy's internal, slow `.copy()` function to straighten out the data and make it contiguous. No matter what you do, that function will need to be called, either by you, or by any imaging library that you pass the `ndarray` into.
- If you like the channel order (such as RGB) but you just want to make the data contiguous, use `img = np.ascontiguousarray(img)`. It will automatically make a copy if the data wasn't already contiguous: https://docs.scipy.org/doc/numpy/reference/generated/numpy.ascontiguousarray.html
- If you want to change the channel order (such as RGB to BGR, or whatever), use `img = cv2.cvtColor(img, cv2.COLOR_<your desired conversion>)`, which will internally do the `.copy` and will then rapidly convert the colors for you all in one command.

**Comment by VideoPlayerCode** [comment]
By the way, matplotlib is affected (slowed down) too, but not by as much, because I looked at their source code and saw that they read the "strides" of Numpy and loop over it manually, thus addressing the data as-is: https://github.com/matplotlib/matplotlib/blob/master/src/numpy_cpp.h

But despite being able to read non-contiguous data, such data slows down matplotlib too:

---

**1080p Non-Contiguous:**

```
C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([1080,1920,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100; x = x[...,::-1]" "plt.imshow(x)"
1 loop, best of 5: 33.5 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([1080,1920,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100; x = x[...,::-1]" "plt.imshow(x)"
10 loops, best of 5: 33.8 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([1080,1920,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100; x = x[...,::-1]" "plt.imshow(x)"
10 loops, best of 5: 33.9 msec per loop
```

Average: 33.73 ms.

---

**1080p Contiguous:**

```
C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([1080,1920,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100" "plt.imshow(x)"
10 loops, best of 5: 27.4 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([1080,1920,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100" "plt.imshow(x)"
10 loops, best of 5: 27.4 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([1080,1920,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100" "plt.imshow(x)"
10 loops, best of 5: 27.7 msec per loop
```

Average: 27.50 ms.

---

**4K Non-Contiguous:**

```
C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([2160,3840,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100; x = x[...,::-1]" "plt.imshow(x)"
1 loop, best of 5: 134 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([2160,3840,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100; x = x[...,::-1]" "plt.imshow(x)"
1 loop, best of 5: 132 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([2160,3840,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100; x = x[...,::-1]" "plt.imshow(x)"
1 loop, best of 5: 133 msec per loop
```

Average: 133.00 ms.

---

**4K Contiguous:**

```
C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([2160,3840,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100" "plt.imshow(x)"
1 loop, best of 5: 107 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([2160,3840,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100" "plt.imshow(x)"
1 loop, best of 5: 107 msec per loop

C:\Users\Johnny>python -m timeit -s "import numpy as np; import matplotlib.pyplot as plt; x = np.zeros([2160,3840,3], np.uint8); x[:,:,2] = 255; x[:,:,1] = 100" "plt.imshow(x)"
1 loop, best of 5: 107 msec per loop
```

Average: 107.00 ms.

---

So, giving matplotlib non-contiguous data is:
- `26.0 ms slower @ 4K`
- `6.23 ms slower @ 1080p`.

Now compare that to the times it takes to convert contiguous data to a different channel order via `cv2.cvtColor`, which preserves the contiguous order (look at the end of the previous post for these numbers):
- `cv2.cvtColor` takes `5.38 ms @ 4K`.
- `cv2.cvtColor` takes `1.58 ms @ 4K`.

So even when Matplotlib is the target of the data, it's way faster to invoke `cv2.cvtColor` instead of doing the array manipulation tricks.

In short: Always use `cv2.cvtColor`. Or, if data is already in the correct colorspace but isn't always contiguous, use `np.ascontiguousarray`. Both are described at the bottom of the previous post.

I hope this helps! :-)
